Bluetooth: GATT: Add function to get UUIDs for adv data

Thalley opened the pull request:
Add a function that can help an application get get a
partial or complete list of UUIDs for advertisement purposes.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

fixes https://github.com/zephyrproject-rtos/zephyr/issues/33777

Thalley committed:
Bluetooth: GATT: Add function to get UUIDs for adv data

Add a function that can help an application get get a
partial or complete list of UUIDs for advertisement purposes.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented on samples/bluetooth/peripheral/src/main.c:
```diff
@@ -265,6 +262,33 @@ static void bt_ready(void)
 		settings_load();
 	}
 
+	err = bt_gatt_get_svc_uuid_data(BT_DATA_UUID16_ALL, &ad[1],
+					uuid16_buf, sizeof(uuid16_buf));
+	if (err != 0) {
+		printk("Could not encode all UUID16 data: %d\n", err);
+
+		err = bt_gatt_get_svc_uuid_data(BT_DATA_UUID16_SOME, &ad[1],
+						uuid16_buf, sizeof(uuid16_buf));
+		if (err != 0) {
+			printk("Could not encode some UUID16 data: %d\n", err);
+			return;
+		}
+	}
+
+	err = bt_gatt_get_svc_uuid_data(BT_DATA_UUID128_ALL, &ad[2],
+					uuid128_buf, sizeof(uuid128_buf));
+	if (err != 0) {
+		printk("Could not encode UUID128 data: %d\n", err);
+
+		err = bt_gatt_get_svc_uuid_data(BT_DATA_UUID128_SOME, &ad[2],
+						uuid128_buf,
+						sizeof(uuid128_buf));
+		if (err != 0) {
+			printk("Could not encode some UUID16 data: %d\n", err);
+			return;
+		}
+	}
```
An alternative I considered was to automatically change the type to _SOME instead of _ALL if it did not fit, instead of returning an error and letting the application decided what to do.

alwa-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5696,3 +5697,153 @@ void bt_gatt_disconnected(struct bt_conn *conn)
+/* This is an array of UUID values that we return via the
+ * bt_gatt_get_svc_uuid_data for advertisements
+ */
+static uint16_t svc_uuid16_data_skip_list[] = {
```
I suggest "exclude list" instead of "skip list". "skip list" can be confused with the type of data structure with that name.

alwa-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5696,3 +5697,153 @@ void bt_gatt_disconnected(struct bt_conn *conn)
+	bool truncated;
+};
+
+/* This is an array of UUID values that we return via the
```
I think this should be "... that we don't return ...", right?

We should also add info about the selection criteria for the entries of this list. I suggest "List of Service UUIDs that are mandatory on a GATT server. Adverts for these Service UUIDs are redundant. bt_gatt_get_svc_uuid_data() filters out entries that are on this list."

alwa-nordic commented on include/bluetooth/gatt.h:
```diff
@@ -1773,6 +1773,35 @@ int bt_gatt_unsubscribe(struct bt_conn *conn,
 void bt_gatt_cancel(struct bt_conn *conn, void *params);
 
+/** @brief Encode GATT UUID of current services into a bt_data struct.
+ *
```
We should mention here that the mandatory GATT services are not included.

alwa-nordic commented on include/bluetooth/gatt.h:
```diff
@@ -1773,6 +1773,38 @@ int bt_gatt_unsubscribe(struct bt_conn *conn,
+ *
+ * @retval -EINVAL if any of the parameters are invalid (e.g. NULL or 0).
+ *
+ * @retval -ENOMEM if a _ALL type is used and @p data_buf does not fit all the
```
The implementation leaves the out-parameters as if a _SOME type was used in the case or -ENOMEM. I can see users relying on this undocumented behavior. I'm not sure if it is a problem. (Is it?) We could deal with it now and specify this behavior in the API, or sabotage the output with memset (or at least set `data_len` to 0).

Thalley replied:
Done :)

Thalley committed:
Bluetooth: GATT: Document attr->user_data types for bt_gatt_attr_func_t

The attr-user_data for bt_gatt_attr_func_t were undocumented
and not very easy to determine if you don't know where to look.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

alwa-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5696,3 +5697,155 @@ void bt_gatt_disconnected(struct bt_conn *conn)
+		return -EINVAL;
+	}
+
+	data->type = type;
```
This is a question, not a suggestion.. Should `data` be initialized / zeroed out here?

Thalley replied:
Not sure I completely follow. 

`bt_data` is a representation of a LTV structure. Resetting it here would not make much sense as we are overwriting all 3 values in the struct already.

alwa-nordic commented on include/bluetooth/gatt.h:
```diff
@@ -393,7 +393,18 @@ enum {
+ *   - char * when UUID is @ref BT_UUID_GATT_CUD
+ *   - @ref bt_gatt_cpf when UUID is @ref BT_UUID_GATT_CPF
+ *
+ *  @param attr Attribute found. The
```
```suggestion
 *  @param attr Attribute found.
```

asbjornsabo commented on samples/bluetooth/peripheral/src/main.c:
```diff
@@ -265,6 +262,33 @@ static void bt_ready(void)
+		err = bt_gatt_get_svc_uuid_data(BT_DATA_UUID16_SOME, &ad[1],
+						uuid16_buf, sizeof(uuid16_buf));
+		if (err != 0) {
+			printk("Could not encode some UUID16 data: %d\n", err);
```
"some" -> "any"?
Either able to encode all data, enable to encode some data, or not able to encode any data.
(The "some" is a bit tricky, if not able to encode all data, some data could not be encoded.  So "could not encode all" implies "could not encode some".)

Also elsewhere.

Thalley replied:
I agree with your statement :) The reason behind using "some" here was that it was attempting to encode "some" (BT_DATA_UUID16_SOME) UUIDs. 

I can change it if you still feel like "any" is better

Thalley replied:
Changed to `Could not encode UUID16 data`, as we at this point won't have any encoded data.

asbjornsabo commented on samples/bluetooth/peripheral/prj.conf:
```diff
@@ -16,10 +16,11 @@ CONFIG_BT_DEVICE_APPEARANCE=833
-CONFIG_FLASH_MAP=y
-CONFIG_NVS=y
-CONFIG_SETTINGS=y
+# CONFIG_BT_KEYS_OVERWRITE_OLDEST=y
```
The changes to this file seem unrelated to the rest of the commit.  Are they intended?

Thalley commented:
Rebased to solve merge conflicts

MariuszSkamra reviewed (changes requested):
I am not convinced this is right thing to do.

Most of the GATT Services don't need to to be advertised. There might be many circumstances under which service should not be advertised e.g. depending on privacy, whether the advertisement is connectable or not etc.
My feeling is that should be left to the user to build advertising data (like it is now).
Alternatively, `bt_gatt_service` could be extended with some callbacks that are called by some advertisement data builder, so that the service could decide whether it should be included in advertisement or not (depending on advertising type, privacy, etc).

Thalley commented:
> 



> I am not convinced this is right thing to do.
> 
> Most of the GATT Services don't need to to be advertised. There might be many circumstances under which service should not be advertised e.g. depending on privacy, whether the advertisement is connectable or not etc. My feeling is that should be left to the user to build advertising data (like it is now). 

I don't disagree with this, which is why this is just a helper function to retrieve that data, and then it's up to the application to determine whether to include it all, or some. However, perhaps this should return an array of UUIDs instead of the data preformatted for adv data? 

> Alternatively, `bt_gatt_service` could be extended with some callbacks that are called by some advertisement data builder, so that the service could decide whether it should be included in advertisement or not (depending on advertising type, privacy, etc).

That seems to be the opposite of what you write above, isn't it? IIRC there is no requirement for any service to have it's UUID advertised. Some services/profile may, however, require that _service data_ (which is different from this) shall be advertised. For _service data_ I could see a benefit of such a callback.

Thalley commented:
Rebased to solve merge conflict

carlescufi commented:
@MariuszSkamra @asbjornsabo please re-review
@jori-nordic @alwa-nordic @MarekPieta @KAGA164 @jhedberg can you please take a look?

jhedberg commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -1817,6 +1828,39 @@ int bt_gatt_unsubscribe(struct bt_conn *conn,
+ * @retval -ENOMEM if a _ALL type is used and @p data_buf does not fit all the
+ *         data. In this case @p data_buf and @p data will be invalid.
+ */
+int bt_gatt_get_svc_uuid_data(uint8_t type, struct bt_data *data,
+			      uint8_t data_buf[], size_t data_buf_size);
```
Couldn't this API be simplified to have just two input parameters, if you require the caller to initialize the `bt_data` struct up-front with the type and buf values? Then the only additional thing you need to provide is the maximum length (i.e. size of the buffer).

You also either need to fix the documentation or the API name to use the same verb. In the documentation you say "encode" but in the function name you say "get".

jhedberg replied:
Just to add to this, if you insist on keeping the `type` and `data_buf` separate and only assigned to `bt_data` inside the call, then I would at least order the parameters so that `bt_data` becomes the first parameter,  Since that's really the only "out" parameter (I'm not sure it's right to call `data_buf` "out", since I don't think the caller is supposed to directly touch its content after the call, rather the "end product" of the call is the `bt_data` struct.

Thalley replied:
I agree with your comments, and we can require that the user initializes the `bt_data` first with the `type` and `data` pointer. 

I'll try to make this change in a new commit (so it's easier to revert), and then I can just squash it if that's what we want.

Thalley replied:
@jhedberg Should I revert the commit to my original proposed API, or should we do something else? 

We could also, instead of using `struct bt_data`, simply put all UUID data in a `uint8_t` buffer, and then let the let the user put that buffer into their `struct bt_data`.

jhedberg replied:
Hmm, that's a tricky one. The reason it's const is to allow assignments to it from runtime evaluated literal values (like most samples are now doing).

> We could also, instead of using `struct bt_data`, simply put all UUID data in a `uint8_t` buffer, and then let the let the user put that buffer into their `struct bt_data`.

Yeah, that might be a cleaner solution, although the encoding format really only has a single user (advertising data). That said, the format is not special - just an array of consecutive values.

jhedberg replied:
Btw, if we go with this "encode into uint8_t array" solution  we also need to let the caller know if everything fit or not, and then let the caller then decide the advertising data type value. Or to put it in other words, when we remove the bt_data type from the API, it should not use BT_DATA_UUID* values as either input or output, to make it generic in clean way.

jhedberg replied:
...it might be possible to convince me otherwise on that last point.. just thinking out loud. Let me know what you think

Thalley replied:
I'm also just thinking aloud here:

```c
 * @retval Number of octets put in @p data_buf.
 *
 * @retval -EINVAL if any of the parameters are invalid (e.g. NULL or 0).
 *         In this case @p data_buf and @p data will be invalid.
 *
 * @retval -ENOMEM if not all service UUIDs could be placed in @p data_buf. 
 *         @p data_buf will contain as many UUIDs as could be placed. In this 
 *         case @p data_buf can still be used with the BT_DATA_UUID16_SOME, 
 *         BT_DATA_UUID32_SOME or BT_DATA_UUID128_SOME depending on the @p type.
 */
int bt_gatt_get_svc_uuid_data(uint8_t type, uint8_t data_buf, size_t data_buf_size);
```

and where `type` would be one of BT_UUID_TYPE_16, BT_UUID_TYPE_32, BT_UUID_TYPE_128.

i.e. the function will pack as many service UUIDs in the `data_buf` as possible, and return the number of octets put in it. 

The returned buffer and size should be enough to then encode the `struct bt_data`. 

But like others, I do have doubt about the usefulness of this API, as A) it's pretty trivial to encode the service UUIDs yourself today, and it's quite difficult to provide a good API for this.

Thalley replied:
@jhedberg @MariuszSkamra What do you think? Is it worth pursuing this API, or just leave it?

jhedberg replied:
I'm ok leaving it. We can always revisit it later in the future if necessary (e.g. new use cases or users come along)

Thalley replied:
@jhedberg Should we leave it open, move it to draft or close it then? 

If we are not planning on merging it anytime soon, I'm in favor of just closing it.

jhedberg replied:
Closing is probably the best option, but maybe keep the branch in your fork so that this can be revisited later.

jhedberg commented on samples/bluetooth/peripheral/src/main.c:
```diff
@@ -287,6 +284,33 @@ static void bt_ready(void)
 		settings_load();
 	}
 
+	err = bt_gatt_get_svc_uuid_data(BT_DATA_UUID16_ALL, &ad[1],
+					uuid16_buf, sizeof(uuid16_buf));
+	if (err != 0) {
+		printk("Could not encode all UUID16 data: %d\n", err);
+
+		err = bt_gatt_get_svc_uuid_data(BT_DATA_UUID16_SOME, &ad[1],
+						uuid16_buf, sizeof(uuid16_buf));
+		if (err != 0) {
+			printk("Could not encode UUID16 data: %d\n", err);
+			return;
+		}
+	}
+
+	err = bt_gatt_get_svc_uuid_data(BT_DATA_UUID128_ALL, &ad[2],
+					uuid128_buf, sizeof(uuid128_buf));
+	if (err != 0) {
+		printk("Could not encode UUID128 data: %d\n", err);
+
+		err = bt_gatt_get_svc_uuid_data(BT_DATA_UUID128_SOME, &ad[2],
+						uuid128_buf,
+						sizeof(uuid128_buf));
+		if (err != 0) {
+			printk("Could not encode UUID128 data: %d\n", err);
+			return;
+		}
+	}
+
```
Is this really a good sample to choose to demonstrate the API? You end up adding many more lines and complexity compared to the original code. In fact, do you actually have a good use case for this API, where you end up with more readable code after the change?

Thalley replied:
Yes, I have a better sample (unicast_audio_server) in mind that I can use instead :)

jhedberg reviewed (changes requested):
The general question I end up asking myself is that does this API really provide some added value over statically defining the advertising data (like the sample you're modifying used to do). I could imagine there maybe being some value for super complex and dynamic applications where the set of advertised services changes, and where the services are split into separate modules that you don't have direct visibility to (so querying the GATT layer that centrally manages these makes sense), but then we'd need to show that in the form of some sample application, rather than modifying one where this API doesn't really make the code any better (rather the opposite, IMO).

asbjornsabo reviewed (dismissed):
My comments have been addressed

Thalley committed:
Bluetooth: GATT: Modify bt_gatt_get_svc_uuid_data to fewer args

Modified the argument list of bt_gatt_get_svc_uuid_data so
that it takes just a struct bt_data and a buffer size,
making the API simpler.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
Closing based on feedback. The idea may be revisited at some later point if we have more requests for it.
